lib/repl.js

```diff
@@ -314,40 +308,42 @@ function REPLServer(prompt,
       debug('finish', e, ret);
       self.memory(cmd);
 
-      if (e && !self.bufferedCommand && cmd.trim().match(/^npm /)) {
-        self.outputStream.write('npm should be run outside of the ' +
-                                'node repl, in your normal shell.\n' +
-                                '(Press Control-D to exit.)\n');
-        self.bufferedCommand = '';
-        self.displayPrompt();
-        return;
-      }
-
       // If error was SyntaxError and not JSON.parse error
       if (e) {
-        if (e instanceof Recoverable) {
+        if (!self.bufferedCommand && cmd.trim().match(/^npm /)) {
+          self.outputStream.write('npm should be run outside of the ' +
+                                  'node repl, in your normal shell.\n' +
+                                  '(Press Control-D to exit.)\n');
+          self.bufferedCommand = '';
+          self.displayPrompt();
+        } else if (e instanceof Recoverable) {
           // Start buffering data like that:
           // {
           // ...  x: 1
           // ... }
           self.bufferedCommand += cmd + '\n';
           self.displayPrompt();
-          return;
         } else {
-          self._domain.emit('error', e);
+          debug('not recoverable, send to domain');
+          process.domain.emit('error', e);
         }
+        return;
       }
 
-      // Clear buffer if no SyntaxErrors
-      self.bufferedCommand = '';
-
-      // If we got any output - print it (if no error)
-      if (!e && (!self.ignoreUndefined || ret !== undefined)) {
+      // If we got any output - print it
+      if (!self.ignoreUndefined || ret !== undefined) {
         self.context._ = ret;
-        self.outputStream.write(self.writer(ret) + '\n');
+        try {
+          var s = self.writer(ret);
+        } catch(e) {
+          process.domain.emit('error', e);
+          return;
+        }
+        self.outputStream.write(s + '\n');
       }
 
       // Display prompt again
+      self.bufferedCommand = '';
```

Comment:
In every other case a modifying `self.bufferedCommand` is grouped with `self.displayPrompt`. Do the same here. Having `self.bufferedCommand` run before calling `self.writer` implies that it is necessary for maintaining the integrity of our state, since we could shortcircuit before `self.displayPrompt`. This is not the case since the domain error handler will reset `self.bufferedCommand`.